library/core/src/alloc/layout.rs

```diff
@@ -85,6 +91,17 @@ impl Layout {
         unsafe { Ok(Layout::from_size_align_unchecked(size, align)) }
     }
 
+    /// Internal helper constructor to skip revalidating alignment validity.
+    #[inline]
+    const fn from_size_valid_align(size: usize, align: ValidAlign) -> Result<Self, LayoutError> {
+        // See above for the correctness of this check.
+        if size > isize::MAX as usize - (align.as_nonzero().get() - 1) {
+            return Err(LayoutError);
+        }
+        // SAFTEY: as above, this check is sufficient.
```

Comment:
pondering: does this need to be duplicated?

Could `from_size_align` become something like
```rust
pub const fn from_size_align(size: usize, align: usize) -> Result<Self, LayoutError> {
    let align = ValidAlign::try_from(align).map_err(|_| LayoutError)?;
    from_size_valid_align(size, align)
}
```
to avoid the repetition?

Reply:
That would be *correct*, but for the time being I'm erring on the side of making small changes so there's a chance of saying whether a change is positive or not.